File: 4esp/Codici python-20250530/Open_Merge_stats.py
# -*- coding: utf-8 -*-
"""
Created on Wed May 14 18:17:38 2025

@author: Erika Brattich (R) & Andrea Faggi (python)
"""

#%% libraries

import numpy as np
import pandas as pd
import os

#%% open all

os.chdir('/home/federico/unibo/atmLab/4esp/foglio_verticale_2/')


for i in range(534):
    b = pd.read_fwf('Rec-Vert_2-000001_'+str(i)+' - Stats.txt', widths=[31]*5)
    b.set_index(b['Statistic [units]'], inplace=True, drop=True)

    b = pd.concat([b.loc['Mean [C]'],
                   b.loc['Std. Dev. [C]'],
                   b.loc['Center [C]'],
                   b.loc['Maximum [C]'],
                   b.loc['Minimum [C]'],
                   b.loc['Number of Pixels']], axis=1)
    b = b.T
    b = b.iloc[:,1:5]
    
    if i==0:
        image_temp = pd.DataFrame([b.loc['Mean [C]']['Image'],
                                   b.loc['Std. Dev. [C]']['Image'],
                                   b.loc['Maximum [C]']['Image'][-5:],
                                   b.loc['Minimum [C]']['Image'][-5:]],
                                   index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        box1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 1'],
                                  b.loc['Std. Dev. [C]']['Box 1'],
                                  b.loc['Maximum [C]']['Box 1'][-5:],
                                  b.loc['Minimum [C]']['Box 1'][-5:]],
                                  index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        line1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 2'],
                                   b.loc['Std. Dev. [C]']['Box 2'],
                                   b.loc['Maximum [C]']['Box 2'][-5:],
                                   b.loc['Minimum [C]']['Box 2'][-5:]],
                                   index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        cursor1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 3'],
                                     b.loc['Std. Dev. [C]']['Box 3'],
                                     b.loc['Maximum [C]']['Box 3'][-5:],
                                     b.loc['Minimum [C]']['Box 3'][-5:]],
                                     index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])

        image_sr = image_temp.T
        box1_sr = box1_temp.T 
        line1_sr = line1_temp.T
        cursor1_sr = cursor1_temp.T
 
        
    else:
        image_temp = pd.DataFrame([b.loc['Mean [C]']['Image'],
                                   b.loc['Std. Dev. [C]']['Image'],
                                   b.loc['Maximum [C]']['Image'][-5:],
                                   b.loc['Minimum [C]']['Image'][-5:]],
                                   index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        box1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 1'],
                                  b.loc['Std. Dev. [C]']['Box 1'],
                                  b.loc['Maximum [C]']['Box 1'][-5:],
                                  b.loc['Minimum [C]']['Box 1'][-5:]],
                                  index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        line1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 2'],
                                   b.loc['Std. Dev. [C]']['Box 2'],
                                   b.loc['Maximum [C]']['Box 2'][-5:],
                                   b.loc['Minimum [C]']['Box 2'][-5:]],
                                   index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        cursor1_temp = pd.DataFrame([b.loc['Mean [C]']['Box 3'],
                                     b.loc['Std. Dev. [C]']['Box 3'],
                                     b.loc['Maximum [C]']['Box 3'][-5:],
                                     b.loc['Minimum [C]']['Box 3'][-5:]],
                                     index=['Mean [C]', 'Std. Dev. [C]', 'Maximum [C]', 'Minimum [C]'])
        
        image_temp = image_temp.T
        box1_temp = box1_temp.T
        line1_temp = line1_temp.T       
        cursor1_temp = cursor1_temp.T
               
        image_sr = pd.concat([image_sr, image_temp], ignore_index=True)
        box1_sr = pd.concat([box1_sr, box1_temp], ignore_index=True)
        line1_sr = pd.concat([line1_sr, line1_temp], ignore_index=True)
        cursor1_sr = pd.concat([cursor1_sr, cursor1_temp], ignore_index=True)


del image_temp, box1_temp, line1_temp, cursor1_temp, i

image_sr = image_sr.astype(float)
box1_sr = box1_sr.astype(float)
line1_sr = line1_sr.astype(float)
cursor1_sr = cursor1_sr.astype(float)

# save the dataframe as txt
DIR = '/home/federico/unibo/atmLab/4esp/'
image_sr.to_csv(DIR+'stats_image_vert.txt', sep=',', index=False)
box1_sr.to_csv(DIR+'stats_box1_vert.txt', sep=',', index=False)
line1_sr.to_csv(DIR+'stats_box2_vert.txt', sep=',', index=False)
cursor1_sr.to_csv(DIR+'stats_box3_vert.txt', sep=',', index=False)

File: 4esp/codice R-20250530/Open_Merge_Stats.R
# -*- coding: utf-8 -*-
#Created on Wed May 14 18:17:38 2025
#@authors: Andrea Faggi (Python) & Erika Brattich (R)

#clean environment first
rm(list=ls())

# Load required libraries
library(readr)
library(dplyr)
library(tidyr)


# Set working directory
setwd('C:/Users/erika.brattich2/Downloads/prova/stats/')

# Initialize empty dataframes
image_sr <- NULL
box1_sr <- NULL
line1_sr <- NULL
cursor1_sr <- NULL

# Loop through files (frames): pay attention to their number and eventually change it
for (i in 0:247) {
  # Read fixed-width file
  b <- read_fwf(paste0('Rec-Lab_prova-000001_', i, ' - Stats.txt'), 
                col_positions = fwf_widths(rep(31, 5)))
  
  # Rename columns; the names have to change depending on your choice of cursors, boxes, ...
  colnames(b) <- c('Statistic', 'Image', 'Box 1', 'Line 1', 'Cursor 1')
  
  # Filter and transpose specific rows
  b_filtered <- b %>%
    filter(Statistic %in% c('Mean [C]', 'Std. Dev. [C]', 'Center [C]', 
                            'Maximum [C]', 'Minimum [C]', 'Number of Pixels'))
  
  # Extract specific statistics
  image_temp <- data.frame(
    Mean = b_filtered$Image[b_filtered$Statistic == 'Mean [C]'],
    StdDev = b_filtered$Image[b_filtered$Statistic == 'Std. Dev. [C]'],
    Max = substr(b_filtered$Image[b_filtered$Statistic == 'Maximum [C]'], nchar(b_filtered$Image[b_filtered$Statistic == 'Maximum [C]'])-4, nchar(b_filtered$Image[b_filtered$Statistic == 'Maximum [C]'])),
    Min = substr(b_filtered$Image[b_filtered$Statistic == 'Minimum [C]'], nchar(b_filtered$Image[b_filtered$Statistic == 'Minimum [C]'])-4, nchar(b_filtered$Image[b_filtered$Statistic == 'Minimum [C]']))
  )
  
  box1_temp <- data.frame(
    Mean = b_filtered$`Box 1`[b_filtered$Statistic == 'Mean [C]'],
    StdDev = b_filtered$`Box 1`[b_filtered$Statistic == 'Std. Dev. [C]'],
    Max = substr(b_filtered$`Box 1`[b_filtered$Statistic == 'Maximum [C]'], nchar(b_filtered$`Box 1`[b_filtered$Statistic == 'Maximum [C]'])-4, nchar(b_filtered$`Box 1`[b_filtered$Statistic == 'Maximum [C]'])),
    Min = substr(b_filtered$`Box 1`[b_filtered$Statistic == 'Minimum [C]'], nchar(b_filtered$`Box 1`[b_filtered$Statistic == 'Minimum [C]'])-4, nchar(b_filtered$`Box 1`[b_filtered$Statistic == 'Minimum [C]']))
  )
  
  line1_temp <- data.frame(
    Mean = b_filtered$`Line 1`[b_filtered$Statistic == 'Mean [C]'],
    StdDev = b_filtered$`Line 1`[b_filtered$Statistic == 'Std. Dev. [C]'],
    Max = substr(b_filtered$`Line 1`[b_filtered$Statistic == 'Maximum [C]'], nchar(b_filtered$`Line 1`[b_filtered$Statistic == 'Maximum [C]'])-4, nchar(b_filtered$`Line 1`[b_filtered$Statistic == 'Maximum [C]'])),
    Min = substr(b_filtered$`Line 1`[b_filtered$Statistic == 'Minimum [C]'], nchar(b_filtered$`Line 1`[b_filtered$Statistic == 'Minimum [C]'])-4, nchar(b_filtered$`Line 1`[b_filtered$Statistic == 'Minimum [C]']))
  )
  
  cursor1_temp <- data.frame(
    Mean = b_filtered$`Cursor 1`[b_filtered$Statistic == 'Mean [C]'],
    StdDev = b_filtered$`Cursor 1`[b_filtered$Statistic == 'Std. Dev. [C]'],
    Max = substr(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Maximum [C]'], nchar(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Maximum [C]'])-4, nchar(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Maximum [C]'])),
    Min = substr(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Minimum [C]'], nchar(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Minimum [C]'])-4, nchar(b_filtered$`Cursor 1`[b_filtered$Statistic == 'Minimum [C]']))
  )
  
  # Combine results
  if (is.null(image_sr)) {
    image_sr <- image_temp
    box1_sr <- box1_temp
    line1_sr <- line1_temp
    cursor1_sr <- cursor1_temp
  } else {
    image_sr <- rbind(image_sr, image_temp)
    box1_sr <- rbind(box1_sr, box1_temp)
    line1_sr <- rbind(line1_sr, line1_temp)
    cursor1_sr <- rbind(cursor1_sr, cursor1_temp)
  }
}

# Convert to numeric
image_sr[] <- lapply(image_sr, as.numeric)
box1_sr[] <- lapply(box1_sr, as.numeric)
line1_sr[] <- lapply(line1_sr, as.numeric)
cursor1_sr[] <- lapply(cursor1_sr, as.numeric)

#add a column index, corresponding to the frame - comment if you do not need it
image_sr$ID <- seq.int(nrow(image_sr))
box1_sr$ID <- seq.int(nrow(box1_sr))
line1_sr$ID <- seq.int(nrow(line1_sr))
cursor1_sr$ID <- seq.int(nrow(cursor1_sr))


setwd('C:/Users/erika.brattich2/Downloads/prova/')

# Save dataframes as csv (change to write.table if you prefer other types of txt files)
write.csv(image_sr, 'stats_image.csv', row.names = FALSE)
write.csv(box1_sr, 'stats_box1.csv', row.names = FALSE)
write.csv(line1_sr, 'stats_line1.csv', row.names = FALSE)
write.csv(cursor1_sr, 'stats_cursor1.csv', row.names = FALSE)
